// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: example/todo/todo_service.proto

package todo

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on Todo with the rules defined in the proto
// definition for this message. If any rules are violated, the first error
// encountered is returned, or nil if there are no violations.
func (m *Todo) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on Todo with the rules defined in the
// proto definition for this message. If any rules are violated, the result is
// a list of violation errors wrapped in TodoMultiError, or nil if none found.
func (m *Todo) ValidateAll() error {
	return m.validate(true)
}

func (m *Todo) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	// no validation rules for Title

	// no validation rules for Description

	// no validation rules for Completed

	// no validation rules for Name

	// no validation rules for Age

	if len(errors) > 0 {
		return TodoMultiError(errors)
	}

	return nil
}

// TodoMultiError is an error wrapping multiple validation errors returned by
// Todo.ValidateAll() if the designated constraints aren't met.
type TodoMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m TodoMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m TodoMultiError) AllErrors() []error { return m }

// TodoValidationError is the validation error returned by Todo.Validate if the
// designated constraints aren't met.
type TodoValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e TodoValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e TodoValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e TodoValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e TodoValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e TodoValidationError) ErrorName() string { return "TodoValidationError" }

// Error satisfies the builtin error interface
func (e TodoValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sTodo.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = TodoValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = TodoValidationError{}

// Validate checks the field values on CreateTodoRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *CreateTodoRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on CreateTodoRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// CreateTodoRequestMultiError, or nil if none found.
func (m *CreateTodoRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *CreateTodoRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Title

	// no validation rules for Description

	if len(errors) > 0 {
		return CreateTodoRequestMultiError(errors)
	}

	return nil
}

// CreateTodoRequestMultiError is an error wrapping multiple validation errors
// returned by CreateTodoRequest.ValidateAll() if the designated constraints
// aren't met.
type CreateTodoRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m CreateTodoRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m CreateTodoRequestMultiError) AllErrors() []error { return m }

// CreateTodoRequestValidationError is the validation error returned by
// CreateTodoRequest.Validate if the designated constraints aren't met.
type CreateTodoRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CreateTodoRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CreateTodoRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CreateTodoRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CreateTodoRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CreateTodoRequestValidationError) ErrorName() string {
	return "CreateTodoRequestValidationError"
}

// Error satisfies the builtin error interface
func (e CreateTodoRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCreateTodoRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CreateTodoRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CreateTodoRequestValidationError{}

// Validate checks the field values on CreateTodoResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *CreateTodoResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on CreateTodoResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// CreateTodoResponseMultiError, or nil if none found.
func (m *CreateTodoResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *CreateTodoResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetTodo()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, CreateTodoResponseValidationError{
					field:  "Todo",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, CreateTodoResponseValidationError{
					field:  "Todo",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetTodo()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return CreateTodoResponseValidationError{
				field:  "Todo",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return CreateTodoResponseMultiError(errors)
	}

	return nil
}

// CreateTodoResponseMultiError is an error wrapping multiple validation errors
// returned by CreateTodoResponse.ValidateAll() if the designated constraints
// aren't met.
type CreateTodoResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m CreateTodoResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m CreateTodoResponseMultiError) AllErrors() []error { return m }

// CreateTodoResponseValidationError is the validation error returned by
// CreateTodoResponse.Validate if the designated constraints aren't met.
type CreateTodoResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CreateTodoResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CreateTodoResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CreateTodoResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CreateTodoResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CreateTodoResponseValidationError) ErrorName() string {
	return "CreateTodoResponseValidationError"
}

// Error satisfies the builtin error interface
func (e CreateTodoResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCreateTodoResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CreateTodoResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CreateTodoResponseValidationError{}

// Validate checks the field values on ReadTodoRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *ReadTodoRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ReadTodoRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ReadTodoRequestMultiError, or nil if none found.
func (m *ReadTodoRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ReadTodoRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	if len(errors) > 0 {
		return ReadTodoRequestMultiError(errors)
	}

	return nil
}

// ReadTodoRequestMultiError is an error wrapping multiple validation errors
// returned by ReadTodoRequest.ValidateAll() if the designated constraints
// aren't met.
type ReadTodoRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ReadTodoRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ReadTodoRequestMultiError) AllErrors() []error { return m }

// ReadTodoRequestValidationError is the validation error returned by
// ReadTodoRequest.Validate if the designated constraints aren't met.
type ReadTodoRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ReadTodoRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ReadTodoRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ReadTodoRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ReadTodoRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ReadTodoRequestValidationError) ErrorName() string { return "ReadTodoRequestValidationError" }

// Error satisfies the builtin error interface
func (e ReadTodoRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sReadTodoRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ReadTodoRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ReadTodoRequestValidationError{}

// Validate checks the field values on ReadTodoResponse with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *ReadTodoResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ReadTodoResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ReadTodoResponseMultiError, or nil if none found.
func (m *ReadTodoResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ReadTodoResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetTodo()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, ReadTodoResponseValidationError{
					field:  "Todo",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, ReadTodoResponseValidationError{
					field:  "Todo",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetTodo()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return ReadTodoResponseValidationError{
				field:  "Todo",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return ReadTodoResponseMultiError(errors)
	}

	return nil
}

// ReadTodoResponseMultiError is an error wrapping multiple validation errors
// returned by ReadTodoResponse.ValidateAll() if the designated constraints
// aren't met.
type ReadTodoResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ReadTodoResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ReadTodoResponseMultiError) AllErrors() []error { return m }

// ReadTodoResponseValidationError is the validation error returned by
// ReadTodoResponse.Validate if the designated constraints aren't met.
type ReadTodoResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ReadTodoResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ReadTodoResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ReadTodoResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ReadTodoResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ReadTodoResponseValidationError) ErrorName() string { return "ReadTodoResponseValidationError" }

// Error satisfies the builtin error interface
func (e ReadTodoResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sReadTodoResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ReadTodoResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ReadTodoResponseValidationError{}

// Validate checks the field values on UpdateTodoRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *UpdateTodoRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on UpdateTodoRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// UpdateTodoRequestMultiError, or nil if none found.
func (m *UpdateTodoRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *UpdateTodoRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	// no validation rules for Title

	// no validation rules for Description

	// no validation rules for Completed

	if len(errors) > 0 {
		return UpdateTodoRequestMultiError(errors)
	}

	return nil
}

// UpdateTodoRequestMultiError is an error wrapping multiple validation errors
// returned by UpdateTodoRequest.ValidateAll() if the designated constraints
// aren't met.
type UpdateTodoRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UpdateTodoRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UpdateTodoRequestMultiError) AllErrors() []error { return m }

// UpdateTodoRequestValidationError is the validation error returned by
// UpdateTodoRequest.Validate if the designated constraints aren't met.
type UpdateTodoRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UpdateTodoRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UpdateTodoRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UpdateTodoRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UpdateTodoRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UpdateTodoRequestValidationError) ErrorName() string {
	return "UpdateTodoRequestValidationError"
}

// Error satisfies the builtin error interface
func (e UpdateTodoRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUpdateTodoRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UpdateTodoRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UpdateTodoRequestValidationError{}

// Validate checks the field values on UpdateTodoResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *UpdateTodoResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on UpdateTodoResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// UpdateTodoResponseMultiError, or nil if none found.
func (m *UpdateTodoResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *UpdateTodoResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetTodo()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, UpdateTodoResponseValidationError{
					field:  "Todo",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, UpdateTodoResponseValidationError{
					field:  "Todo",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetTodo()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return UpdateTodoResponseValidationError{
				field:  "Todo",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return UpdateTodoResponseMultiError(errors)
	}

	return nil
}

// UpdateTodoResponseMultiError is an error wrapping multiple validation errors
// returned by UpdateTodoResponse.ValidateAll() if the designated constraints
// aren't met.
type UpdateTodoResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UpdateTodoResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UpdateTodoResponseMultiError) AllErrors() []error { return m }

// UpdateTodoResponseValidationError is the validation error returned by
// UpdateTodoResponse.Validate if the designated constraints aren't met.
type UpdateTodoResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UpdateTodoResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UpdateTodoResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UpdateTodoResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UpdateTodoResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UpdateTodoResponseValidationError) ErrorName() string {
	return "UpdateTodoResponseValidationError"
}

// Error satisfies the builtin error interface
func (e UpdateTodoResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUpdateTodoResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UpdateTodoResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UpdateTodoResponseValidationError{}

// Validate checks the field values on DeleteTodoRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *DeleteTodoRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DeleteTodoRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// DeleteTodoRequestMultiError, or nil if none found.
func (m *DeleteTodoRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *DeleteTodoRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	if len(errors) > 0 {
		return DeleteTodoRequestMultiError(errors)
	}

	return nil
}

// DeleteTodoRequestMultiError is an error wrapping multiple validation errors
// returned by DeleteTodoRequest.ValidateAll() if the designated constraints
// aren't met.
type DeleteTodoRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DeleteTodoRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DeleteTodoRequestMultiError) AllErrors() []error { return m }

// DeleteTodoRequestValidationError is the validation error returned by
// DeleteTodoRequest.Validate if the designated constraints aren't met.
type DeleteTodoRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DeleteTodoRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DeleteTodoRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DeleteTodoRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DeleteTodoRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DeleteTodoRequestValidationError) ErrorName() string {
	return "DeleteTodoRequestValidationError"
}

// Error satisfies the builtin error interface
func (e DeleteTodoRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDeleteTodoRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DeleteTodoRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DeleteTodoRequestValidationError{}

// Validate checks the field values on DeleteTodoResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *DeleteTodoResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DeleteTodoResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// DeleteTodoResponseMultiError, or nil if none found.
func (m *DeleteTodoResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *DeleteTodoResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Success

	if len(errors) > 0 {
		return DeleteTodoResponseMultiError(errors)
	}

	return nil
}

// DeleteTodoResponseMultiError is an error wrapping multiple validation errors
// returned by DeleteTodoResponse.ValidateAll() if the designated constraints
// aren't met.
type DeleteTodoResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DeleteTodoResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DeleteTodoResponseMultiError) AllErrors() []error { return m }

// DeleteTodoResponseValidationError is the validation error returned by
// DeleteTodoResponse.Validate if the designated constraints aren't met.
type DeleteTodoResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DeleteTodoResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DeleteTodoResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DeleteTodoResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DeleteTodoResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DeleteTodoResponseValidationError) ErrorName() string {
	return "DeleteTodoResponseValidationError"
}

// Error satisfies the builtin error interface
func (e DeleteTodoResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDeleteTodoResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DeleteTodoResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DeleteTodoResponseValidationError{}
